src/nlp/search.py

```diff
@@ -0,0 +1,366 @@
+                self.faiss_index = faiss.index_cpu_to_gpu_multiple(
+                    self.faiss_gpu_options.resource_vec,
+                    self.faiss_gpu_options.device_vec,
+                    index,
+                    self.faiss_gpu_options.cloner_options,
+                )
+            else:
+                self.faiss_index = index
+        for i in range(0, len(vectors), batch_size):
+            vecs = vectors[i : i + batch_size] if column is None else vectors[i : i + batch_size][column]
+            self.faiss_index.add(vecs)
+
+    def search(self, query: np.array, k=10):
+        """ Find the nearest examples indices to the query.
+
+            Args:
+                `query` (`np.array`): The query as a numpy array.
+                `k` (`int`): The number of examples to retrieve.
+
+            Ouput:
+                `scores` (`List[List[float]`): The retrieval scores of the retrieved examples.
+                `indices` (`List[List[int]]`): The indices of the retrieved examples.
+        """
+        assert len(query.shape) == 1 or (len(query.shape) == 2 and query.shape[0] == 1)
+        queries = query.reshape(1, -1)
+        if not queries.flags.c_contiguous:
+            queries = np.asarray(queries, order="C")
+        scores, indices = self.faiss_index.search(queries, k)
+        return scores[0], indices[0].astype(int)
+
+    def search_batch(self, queries: np.array, k=10):
+        """ Find the nearest examples indices to the queries.
+
+            Args:
+                `queries` (`np.array`): The queries as a numpy array.
+                `k` (`int`): The number of examples to retrieve.
+
+            Ouput:
+                `scores` (`List[List[float]`): The retrieval scores of the retrieved examples per query.
+                `indices` (`List[List[int]]`): The indices of the retrieved examples per query.
+        """
+        assert len(queries.shape) == 2
+        if not queries.flags.c_contiguous:
+            queries = np.asarray(queries, order="C")
+        scores, indices = self.faiss_index.search(queries, k)
+        return scores, indices.astype(int)
+
+
+class IndexableMixin:
+    """Add indexing features to `nlp.Dataset`"""
+
+    def __init__(self):
+        self._indexes: Dict[str, BaseIndex] = {}
+
+    def __getitem__(self, key):
+        raise NotImplementedError
+
+    def is_index_initialized(self, column: str) -> bool:
+        return column in self._indexes
+
+    def _check_index_is_initialized(self, column: str):
+        if not self.is_index_initialized(column):
+            raise MissingIndex(
+                f"Index with column '{column}' not initialized yet. Please make sure that you call `add_vector_index` or `add_text_index` first."
+            )
+
+    def list_indexes(self) -> List[str]:
+        """List the columns/identifiers of all the attached indexes."""
+        return list(self._indexes)
+
+    def add_vector_index(
+        self,
+        column: str,
+        vectors,
```

Comment:
Shall we give this option to supply external vectors?

Not sure I see how this integrates with our usual "all is in the table" approach.

Reply:
I guess we can add it for now as it can be quite convenient, but we shouldn't expect to add more support for external things as it can end up out of scope pretty quickly.